### PR TITLE
Add `accelerate` support for `Whisper`

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

This PR adds `accelerate` support for `Whisper`, this way `Whisper` can be loaded in 8-bit as follows:
```
from transformers import WhisperModel

model = WhisperModel.from_pretrained("openai/whisper-tiny", device_map="auto", load_in_8bit=True)
```

However some slow tests are not passing when running the whole `Whisper` testing suite but they pass if I run them independently one by one - I still need to understand why this is happening
```
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelTest::test_multi_gpu_data_parallel_forward
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_large_batched_generation
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_large_generation
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_large_generation_multilingual
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_large_logits_librispeech
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_small_en_logits_librispeech
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_tiny_en_batched_generation
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_tiny_en_generation
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_tiny_generation
FAILED tests/models/whisper/test_modeling_whisper.py::WhisperModelIntegrationTests::test_tiny_logits_librispeech
```

cc @sgugger @ArthurZucker @ydshieh 

Thanks!

### CODE REVIEW

**Comment by ArthurZucker** [comment]
For the failing tests, Wait a little bit a PR related to
The tests will be merged ☺️

**Comment by younesbelkada** [comment]
Great! I'll try again then once the PR that you have mentioned will be merged

**Review by sgugger — APPROVED** [review]
Thanks for adding support to Whisper! Did you double-check the offload and multigpu tests pass?

**Review by ArthurZucker — APPROVED** [review]
LGTM

File: src/transformers/models/whisper/modeling_whisper.py
Comment (younesbelkada) [reply src/transformers/models/whisper/modeling_whisper.py:449]: Hum I think this won't work since `WhisperEncoderLayer` has a skip connection with a residual `hidden_states`, We can fix it by forcing [this](https://github.com/huggingface/transformers/blob/072dfdaee4deee65328071e92ac48f471a78490c/src/transformers/models/whisper/modeling_whisper.py#L293) variable to be on the same device than the `hidden_states` after the attention layer but I wanted to avoid extra device moving!

**Comment by younesbelkada** [comment]
We now know the root cause of the initial issue explained above after an intense debugging session with @ArthurZucker !
the `test_disk_offload` test first overrides the `ACT2FN[config.activation_function]` variable that is initialized only once by the testing suite when importing the model. Therefore when running the slow tests that does not use `device_map=...` all submodules will be initialized without the `_hf_hook` variable except for the [`activation_fn`](https://github.com/huggingface/transformers/blob/072dfdaee4deee65328071e92ac48f471a78490c/src/transformers/models/whisper/modeling_whisper.py#L269) attribute since it is not re-initialized. A workaround could be to design a proper `tearDown` function for those tests, or to change the slow tests with `device_map=auto` ;) 
cc @sgugger

**Comment by younesbelkada** [comment]
I can confirm all tests are green after applying the fix proposed in #19728 ! Thanks a lot for adding the fix 💪

**Comment by younesbelkada** [comment]
Since https://github.com/huggingface/transformers/pull/19728 has been merged, I will merge this PR once the tests below are green! 🟢

**Pull request merged**
